### PR TITLE
quit wrapping `history`, make it the official API

### PR DESCRIPTION
**Pull request by ryanflorence**
- The history passed in to a router becomes the
  API to transition around the app inside of a
  component or outside

- History exposes the router’s history as a member
  of the component, replacing Navigation and
  IsActive mixins.

- swapped argument order in `redirectTo` to keep
  the api predictable

- renamed `redirectTo` to `replaceState` because
  that’s what it is

### CODE REVIEW

File: docs/Glossary.md [hunk 1]
```diff
@@ -20,9 +20,9 @@ A *component* is a React component class or a string (e.g. "div"). Basically, it
 
 ## EnterHook
 
-    type EnterHook = (nextState: RouterState, redirectTo: RedirectFunction, callback?: Function) => any;
+    type EnterHook = (nextState: RouterState, replaceState: RedirectFunction, callback?: Function) => any;
```
Comment (mjackson) [inline docs/Glossary.md]: :+1:

File: examples/master-detail/app.js [hunk 2]
```diff
@@ -96,7 +96,7 @@ var Contact = React.createClass({
   destroy() {
     var { id } = this.props.params;
     ContactStore.removeContact(id);
-    this.transitionTo('/');
+    this.history.pushState('/');
```
Comment (mjackson) [inline examples/master-detail/app.js]: Should be `pushState(null, '/')`

File: modules/Router.js [hunk 3]
```diff
@@ -62,6 +62,13 @@ const Router = React.createClass({
     });
   },
 
+  componentWillReceiveProps(nextProps) {
+    warning(
```
Comment (mjackson) [inline modules/Router.js]: :+1:

File: examples/master-detail/app.js (hunk 2 above)
Comment (ryanflorence) [inline examples/master-detail/app.js]: good catch, fixed.

**Comment by mjackson** [comment]
Love this. Excellent work! :+1:

**Comment by knowbody** [comment]
:+1:

**Comment by gaearon** [comment]
Neat!

**Pull request merged**